### PR TITLE
OVFImport: support_instance_import_when_boot_disk_spans_mltiple_physical_disks

### PR DESCRIPTION
**Pull request by MahmoudNada0**
OVFImport: support_instance_import_when_boot_disk_spans_mltiple_physical_disks_or_with_LVM

/cc EricEdens
/cc zoran15

### CODE REVIEW

File: cli_tools/common/image/importer/request.go [hunk 1]
```diff
@@ -72,37 +74,38 @@ func (args *ImageImportRequest) checkRequiredArguments() error {
...
+	Source                        Source `name:"source" validate:"required"`
+	StdoutLogsDisabled            bool
+	StorageLocation               string
+	Subnet                        string
+	SysprepWindows                bool
+	Tool                          daisyutils.Tool `name:"tool" validate:"required"`
+	Timeout                       time.Duration   `name:"timeout" validate:"required"`
+	UefiCompatible                bool
+	Zone                          string `name:"zone" validate:"required"`
+	TanslationWorkerAttachedDisks []*compute.AttachedDisk
```
[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply cli_tools/common/image/importer/request.go:108]: I mean here the data disks (non-bootable disks) only, as boot disk is already passed to the translations workflow itself as ("source_disk" var).
do you suggest any other more clearer name?

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go [hunk 2]
```diff
@@ -291,7 +299,7 @@ func (oi *OVFImporter) getMachineType(
 	return machineTypeProvider.GetMachineType()
 }
 
-func (oi *OVFImporter) importDiskFilesToImages() error {
+func (oi *OVFImporter) importDisksVMDKFiles() error {
```
[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:302]: Done.

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go [hunk 3]
```diff
@@ -313,14 +321,129 @@ func (oi *OVFImporter) importDiskFilesToImages() error {
...
+		return err
+	}
+
+	oi.images = []domain.Image{image.NewImage(request.Project, request.ImageName)}
+
+	return nil
+}
+
+func (oi *OVFImporter) createDisksFromVmdkFiles(diskInfos []ovfutils.DiskInfo, dataDiskPrefix string) error {
+	// create disk for each vmdk file
```
[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:420]: Done.

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go [hunk 4]
```diff
@@ -313,14 +321,129 @@ func (oi *OVFImporter) importDiskFilesToImages() error {
...
+	oi.params.OsID = osIDValue
+
 	oi.machineTypeString, err = oi.getMachineType(ovfDescriptor, *oi.params.Project, oi.params.Zone)
 	if err != nil {
 		return err
 	}
 
 	oi.Logger.User(fmt.Sprintf("Will create instance of `%v` machine type.", oi.machineTypeString))
 
-	return oi.importDisks(osIDValue, &diskInfos)
```
[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:323]: The part of code i added here is main consists of 2 main functionalities:
1) create data disks for all VMDK data files.
2) import the boot disk VMDK file into image, it is be called after the data disks are created.
I did't use multiImage importer as I need to create the data disks first then import boot image that will use these disks, what do u think?

File: cli_tools/common/image/importer/request.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (EricEdens) [reply cli_tools/common/image/importer/request.go:108]: Maybe `DataDisks`? Or something like that?

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go [hunk 5]
```diff
@@ -313,14 +321,129 @@ func (oi *OVFImporter) importDiskFilesToImages() error {
...
+	return nil
+}
+
+func (oi *OVFImporter) createDisksFromVmdkFiles(diskInfos []ovfutils.DiskInfo, dataDiskPrefix string) error {
+	// create disk for each vmdk file
+
+	params := *oi.params
+
+	var disks []domain.Disk
+	for i, info := range diskInfos {
```
Comment (EricEdens) [inline cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:425]: It looks like you're creating blank disks here; don't you want to write the OVF file's VMDK contents to the disk?

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (EricEdens) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:323]: I don't think `multiImageImporter` is used anymore after your change, and it looks like the code you've added to this file is conceptually similar to the goal of `multiImageImporter`: Given multiple VMDK files, import them to GCP in parallel.

You'll want to put your new code behind an interface, and call that interface from `ovf_importer.go`. If you want to use `multiImageImporter` as that interface, I'm happy with that. If you want to delete it and create a different interface, that's great too.

The reason for this refactoring is to avoid adding detailed logic to `ovf_importer.go`. This file is really large, so testing detailed logic is difficult given the number of dependencies that need to be constructed. By putting your new code behind a dedicated interface, you have a minimal number of dependencies, and testing becomes easier. Read through the code in `multi_image_importer`; it splits functionality into files such that it's easy to test corner cases using unit tests.

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go [hunk 6]
```diff
@@ -313,14 +321,129 @@ func (oi *OVFImporter) importDiskFilesToImages() error {
...
+	if err != nil {
+		return err
+	}
+
+	oi.images = []domain.Image{image.NewImage(request.Project, request.ImageName)}
+
+	return nil
+}
+
+func (oi *OVFImporter) createDisksFromVmdkFiles(diskInfos []ovfutils.DiskInfo, dataDiskPrefix string) error {
```
[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go]: Now, I reused the inflation interface to create data-disks for OVF import

File: cli_tools/common/disk/disk.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+package disk
+
+import (
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/domain"
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/utils/daisyutils"
+)
+
+type defaultDisk struct {
+	Project, Zone, DiskName, URI string
```
[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply cli_tools/common/disk/disk.go]: Done.

File: cli_tools/common/disk/disk.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+// NewDisk constructs an Disk,
+// but rather a convenience object for passing the disk's project, disk name, disk zone,
+// and URI.
+func NewDisk(project, zone, diskName string) domain.Disk {
+	return &defaultDisk{
```
[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply cli_tools/common/disk/disk.go]: Done

File: cli_tools/common/image/importer/request.go [hunk 9]
```diff
@@ -103,6 +104,8 @@ type ImageImportRequest struct {
 	Timeout               time.Duration   `name:"timeout" validate:"required"`
 	UefiCompatible        bool
 	Zone                  string `name:"zone" validate:"required"`
+	DiskName              string
```
Comment (EricEdens) [inline cli_tools/common/image/importer/request.go:107]: This is creating a leaky abstraction, where the caller needs to understand when to provide `DiskName` and when to provide `ImageName`. My intuition is the safer alternative is to not provide this field, and instead determine the PD's URI by looking at the return value when you call `Inflate`. What do you think?

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go [hunk 10]
```diff
@@ -313,14 +321,129 @@ func (oi *OVFImporter) importDiskFilesToImages() error {
...
+	if len(diskInfos) > 0 {
+		if err = oi.importBootDiskImage(diskInfos[0], disksNamesPrefix); err != nil {
+			return err
+		}
+	}
+
+	return err
+}
+
+func (oi *OVFImporter) importBootDiskImage(bootDiskInfos ovfutils.DiskInfo, disksNamesPrefix string) error {
```
[the remark this answers is not in the document]
Comment (EricEdens) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:363]: On second read, I think it's pretty clear.

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go [hunk 11]
```diff
@@ -313,14 +322,66 @@ func (oi *OVFImporter) importDiskFilesToImages() error {
...
+		}
+		oi.disks = disks
+	}
+
+	if len(diskInfos) > 0 {
+		if err = oi.importBootDiskImage(diskInfos[0], disksNamesPrefix); err != nil {
+			return err
+		}
+	}
+
```
[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:359]: It shouldn't happen, but this's a safe condition just to handle if a corrupted GCS folder that has only ovf file without any vmdk file.

File: cli_tools/common/image/importer/request.go (hunk 9 above)
Comment (MahmoudNada0) [reply cli_tools/common/image/importer/request.go:107]: I thought about it, but I just wanted to save the URIs of the created disks in the same order like the filesURIs and as requests, like what is done in the multi_image_importer before.
I can use the same way of naming disks like inflater (exectionID + idx) [here](https://github.com/GoogleCloudPlatform/compute-image-tools/blob/master/cli_tools/common/image/importer/inflater.go#L309) to get the disk name but I feel like it's not good idea because it will fail if the inflater changed it's default way of naming disks. what do u think?

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go [hunk 12]
```diff
@@ -483,3 +515,41 @@ func (oi *OVFImporter) getOsIDValue(descriptor *ovf.Envelope) (osIDValue string,
 	}
 	return osIDValue, nil
 }
+
+func (oi *OVFImporter) buildBootDiskImageImportRequest(imageName string, bootDiskFilePath string) (request importer.ImageImportRequest, err error) {
```
[the remark this answers is not in the document]
Comment (EricEdens) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:519]: Oh, thanks -- you're right, the new location for `request_builder` is `multi_disk_importer`. The same observation exists -- about 30 lines in `ovf_importer.go` is mostly duplicated in `request_builder`.

File: cli_tools/common/image/importer/request.go (hunk 9 above)
Comment (EricEdens) [reply cli_tools/common/image/importer/request.go:107]: To maintain order, can't you do something like this?

```
	for i, request := range requests {
		req := request
		disk, err := disk.NewDisk(req.Project, req.Zone, req.DiskName)
		if err != nil {
			return nil, err
		}

		disks = append(disks, disk)
		logPrefix := fmt.Sprintf("[import-%s]", req.DaisyLogLinePrefix)

		group.Go(func() error {
                        // Use result of r.singleDiskImporter.Import to update disks[i]
			return r.singleDiskImporter.Import(ctx, req, r.logger.NewLogger(logPrefix))
		})
	}
```

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go [hunk 13]
```diff
@@ -483,3 +515,41 @@ func (oi *OVFImporter) getOsIDValue(descriptor *ovf.Envelope) (osIDValue string,
...
+	bootDiskFilePathSource, err := importer.NewSourceFactory(oi.storageClient).Init(bootDiskFilePath, "")
+
+	if err != nil {
+		return request, err
+	}
+
+	request = importer.ImageImportRequest{
+		ExecutionID:           imageName,
```
Comment (yswe) [inline cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:527]: Do you know why this `ExecutionID` is not generated but just assigned an imageName? In other places it is a build ID, I am a bit confused here.

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go [hunk 14]
```diff
@@ -313,14 +322,66 @@ func (oi *OVFImporter) importDiskFilesToImages() error {
...
+
+func (oi *OVFImporter) importBootDiskImage(bootDiskInfos ovfutils.DiskInfo, disksNamesPrefix string) error {
+
+	oi.Logger.User(fmt.Sprint("Importing boot Disk Image ..."))
+
+	imageName := fmt.Sprintf("%s-%s", disksNamesPrefix, "boot-image")
+	request, err := oi.buildBootDiskImageImportRequest(imageName, bootDiskInfos.FilePath)
+
+	if oi.imageImporter == nil {
+		oi.imageImporter, err = importer.NewImporter(request, oi.computeClient, oi.storageClient, oi.Logger)
```
Comment (yswe) [inline cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:371]: Could you clarify why do we need here a lazy assignment? As I see this method is called only once. Have you checked a way to avoid having a nullable field here?

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (yswe) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:303]: Does this method still work only for VMDK files, or do we need update naming here?

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (yswe) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:359]: I think the condition when `len(diskInfos) == 0` won't be handled by `return err` as the value of `err` will be `nil`. If you consider handling this case, probably you should add some custom error here.

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go [hunk 15]
```diff
@@ -378,13 +423,15 @@ func (oi *OVFImporter) Import() error {
 	if err := oi.paramValidator.ValidateAndPopulate(oi.params); err != nil {
 		return err
 	}
-	if err := oi.importDiskFilesToImages(); err != nil {
-		oi.imageDeleter.DeleteImagesIfExist(oi.images)
+	if err := oi.importDisksVMDKFiles(); err != nil {
+		oi.resourceDeleter.DeleteImagesIfExist(oi.images)
```
Comment (yswe) [inline cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:427]: Could you clarify why do we need to delete images? I thought that they wouldn't be created at all. Do I miss anything?

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go [hunk 16]
```diff
@@ -378,13 +423,15 @@ func (oi *OVFImporter) Import() error {
...
+		oi.resourceDeleter.DeleteDisksIfExist(oi.disks)
 		return err
 	}
 	if err := oi.createWorkerForFinalInstance().Run(map[string]string{}); err != nil {
 		oi.Logger.User(err.Error())
-		oi.imageDeleter.DeleteImagesIfExist(oi.images)
+		oi.resourceDeleter.DeleteImagesIfExist(oi.images)
```
Comment (yswe) [inline cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:433]: Could you clarify why do we need to delete images? I thought that they wouldn't be created at all. Do I miss anything?

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go (hunk 15 above)
Comment (EricEdens) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:427]: good find -- I think this `DeleteImagesIfExist` can be removed here since mahmoud is only making disks at this point.

File: cli_tools/gce_ovf_import/multi_disk_importer/request_executor.go [hunk 17]
```diff
@@ -12,54 +12,59 @@
...
-			return images, daisy.Errf("Intermediate image %s already exists. Re-run import.", request.ImageName)
-		}
-	}
-	for _, request := range requests {
+	for i, request := range requests {
 		req := request
-		images = append(images, image.NewImage(request.Project, request.ImageName))
-		logPrefix := fmt.Sprintf("[import-%s]", req.DaisyLogLinePrefix)
+		diskIdx := i
+		logPrefix := fmt.Sprintf("[import-disk-%d]", i+1)
```
[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply cli_tools/gce_ovf_import/multi_disk_importer/request_executor.go:49]: we don't have the diskName here, the diskURI is returned from singleDiskImporter.Import function and diskName is extracted from this URI. See line 51

File: cli_tools/common/image/importer/inflater.go [hunk 18]
```diff
@@ -310,6 +312,11 @@ func getDiskName(executionID string) string {
 	return fmt.Sprintf("disk-%v", executionID)
 }
 
+// GetDiskURI return the URI of a PD disk
```
[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply cli_tools/common/image/importer/inflater.go:315]: according to golint, any public function should have a comment on it.

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go (hunk 13 above)
Comment (MahmoudNada0) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:527]: The executionID is used as a substring of the disks names that are created in the inflation step, so I think it's make sense to be assigned an imageName, also it was name like this in the old [multiImage-importer](https://github.com/GoogleCloudPlatform/compute-image-tools/blob/master/cli_tools/gce_ovf_import/multi_image_importer/request_builder.go#L41)
but, really i don't know when it should be assigned as BuildID, @EricEdens could you please clarify this part for us?

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:359]: I don't know if we should return error in this case or not. @EricEdens

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go (hunk 16 above)
Comment (MahmoudNada0) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:433]: But still we create image for the boot disk, in case of failure I think we need to delete it.

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:303]: I think it does make sense to change it.

File: daisy_workflows/linux_common/utils/diskutils.py [hunk 19]
```diff
@@ -37,7 +39,20 @@ def log_key_value(key, value):
   print(_STATUS_PREFIX + "<serial-output key:'%s' value:'%s'>" % (key, value))
 
 
-def MountDisk(disk) -> guestfs.GuestFS:
+def get_physical_drives():
```
[the remark this answers is not in the document]
Comment (yswe) [reply daisy_workflows/linux_common/utils/diskutils.py:42]: @MahmoudNada0, in the case of a failure your method won't be consistent in the sense of the returned value:

1. If `lsblk` is completed successfully then you will return all disks including the bootable one;
2. If `lsblk` fails then you will return only a data disk.

As a caller I will never know if the output contains a bootable disk or not because I will never know if an error has happen in `lsblk`.

However, I would agree with @EricEdens here, that if `lsblk` fails then it is not something we should hide, I would fail with it.

[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply daisy_workflows/linux_common/utils/diskutils.py:42]: Ok Good, Then we can instead of returning only the "/dev/sdb" I'll return an error and make the import fail

File: cli_tools/common/image/importer/inflater.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (yswe) [reply cli_tools/common/image/importer/inflater.go:315]: Actually my question is about the body of the method 😃 , however, you have already answered to it in a different thread.

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go (hunk 15 above)
Comment (yswe) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:427]: I am not very familiar with this part of code, therefore it would be good to know for the future, why do we need to create an image for a bootable disk at all to import an instance. May I ask you to clarify it, please?

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go (hunk 14 above)
Comment (yswe) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:371]: This part is slightly not clear to me, the method `importBootDiskImage` accepts two arguments: `bootDiskInfos` and `disksNamesPrefix`.

- `disksNamesPrefix` is quite a straightforward argument which value we can define in the early beginning;
- `bootDiskInfos` is constructed **before** importing disks, and uses only `ovfutils.GetOVFDescriptorAndDiskPaths`. `ofvDescriptorLoader` is created in the importer constructor: https://github.com/GoogleCloudPlatform/compute-image-tools/blob/1675a8ccca681f85284255749386ec6e8d786567/cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go#L123 It looks to me that we can move the building of `imageImporter` there, do I miss anything?

Comment (MahmoudNada0) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:371]: I'm talking about **NewImporter** function not **importBootDiskImage**, NewImporter accepts the ImageImportRequest object as a parameter, and I've to create the dataDisks before building the boot-disk ImageImportRequest as it uses them.

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go (hunk 15 above)
Comment (MahmoudNada0) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:427]: Because we need to import the boot disk using image importer to be passed on al the stages of image import, also to be translated, and this is demonstrated in the design document, am I answered your question?

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go (hunk 14 above)
Comment (yswe) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:371]: Well, in this particular context it means exactly the same:

```go
func (oi *OVFImporter) importBootDiskImage(bootDiskInfos ovfutils.DiskInfo, disksNamesPrefix string) error {

	oi.Logger.User(fmt.Sprint("Importing boot Disk Image ..."))

	imageName := fmt.Sprintf("%s-%s", disksNamesPrefix, "boot-image")
	request, err := oi.buildBootDiskImageImportRequest(imageName, bootDiskInfos.FilePath)

	if oi.imageImporter == nil {
		oi.imageImporter, err = importer.NewImporter(request, oi.computeClient, oi.storageClient, oi.Logger)
		if err != nil {
...
```

To create an image importer you need the following set of arguments: `request`, `oi.computeClient`, `oi.storageClient`, `oi.Logger`.

`oi.computeClient`, `oi.storageClient`, `oi.Logger` are created in the `NewOVFImporter` method, therefore the only argument you need to consider is `request`.

`request` is created by a call to `oi.buildBootDiskImageImportRequest(imageName, bootDiskInfos.FilePath)` that requires `imageName` and `bootDiskInfos.Filepath`.
 - `imageName` is created from `disksNamesPrefix` that is passed to `importBootDiskImage`;
 - `bootDiskInfos` is passed to `importBootDiskImage` as well.

And as I see from the code, you don't have to create data disks to initialise `disksNamesPrefix` and `bootDiskInfos`:

```go
	var disksNamesPrefix string
	if oi.params.IsInstanceImport() {
		disksNamesPrefix = oi.params.InstanceNames
	} else {
		disksNamesPrefix = oi.params.MachineImageName
	}
```

```go
	ovfDescriptor, diskInfos, err := ovfutils.GetOVFDescriptorAndDiskPaths(
		oi.ovfDescriptorLoader, ovfGcsPath)
	if err != nil {
		return err
	}
```

Therefore it looks to me that the importer can be created inside the `NewOVFImporter` method. Do I miss anything?

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go (hunk 15 above)
Comment (yswe) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:427]: I ask about a slightly different thing. I see that you use this image here: https://github.com/GoogleCloudPlatform/compute-image-tools/blob/1675a8ccca681f85284255749386ec6e8d786567/cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go#L389

However, do we really need an image in the end to create an instance? Can we modify the workflow in the way that it doesn't require an image but only a reference to a bootable disk?

@EricEdens, do we have any technical limitations here?

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go (hunk 14 above)
Comment (MahmoudNada0) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:371]: But DataDisks which is part of the request is constructed after calling this function "oi.multiDiskImporter.Import()"

File: cli_tools/gce_ovf_import/multi_disk_importer/request_builder.go [hunk 20]
```diff
@@ -28,47 +29,48 @@ type requestBuilder struct {
 	sourceFactory importer.SourceFactory
 }
 
-// buildRequests constructs a list of ImageImportRequests based on the user's
+// buildRequests constructs a list of requests to import data disks based on the user's
 // invocation parameters and the files referenced in the OVF descriptor.
 func (r *requestBuilder) buildRequests(params *ovfdomain.OVFImportParams, fileURIs []string) (requests []importer.ImageImportRequest, err error) {
-	for i, fileURI := range fileURIs {
+	for i, dataDiskURI := range fileURIs {
```
[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply cli_tools/gce_ovf_import/multi_disk_importer/request_builder.go:35]: Done.

File: cli_tools/common/image/importer/request.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply cli_tools/common/image/importer/request.go]: Done.

File: cli_tools/gce_ovf_import/multi_disk_importer/request_builder.go [hunk 21]
```diff
@@ -28,47 +29,48 @@ type requestBuilder struct {
...
 			Zone:                  params.Zone,
-		}
-		bootable := i == 0
-		if bootable {
-			request.OS = params.OsID
-			request.BYOL = params.BYOL
-			importer.FixBYOLAndOSArguments(&request.OS, &request.BYOL)
-		} else {
-			request.DataDisk = true
+			OS:                    params.OsID,
```
[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply cli_tools/gce_ovf_import/multi_disk_importer/request_builder.go]: Done

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go [hunk 22]
```diff
@@ -483,3 +515,41 @@ func (oi *OVFImporter) getOsIDValue(descriptor *ovf.Envelope) (osIDValue string,
...
+		Project:               *oi.params.Project,
+		ScratchBucketGcsPath:  pathutils.JoinURL(oi.params.ScratchBucketGcsPath, imageName),
+		Source:                bootDiskFilePathSource,
+		StdoutLogsDisabled:    oi.params.StdoutLogsDisabled,
+		Subnet:                oi.params.Subnet,
+		Timeout:               oi.params.Deadline.Sub(time.Now()),
+		Tool:                  oi.params.GetTool(),
+		UefiCompatible:        oi.params.UefiCompatible,
+		Zone:                  oi.params.Zone,
+		CreatedDataDisks:      oi.disks,
```
[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go]: Done

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:519]: We've already discussed this part in a meeting.

File: cli_tools/common/image/importer/request.go (hunk 9 above)
Comment (MahmoudNada0) [reply cli_tools/common/image/importer/request.go]: Done.

File: daisy_workflows/linux_common/utils/diskutils.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply daisy_workflows/linux_common/utils/diskutils.py:42]: As agreed with @yswe we will warn that failed to execute 'lsblk' cmd, and continue anyway assuming that there are two disks only ['/dev/sda', '/dev/sdb']

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go (hunk 14 above)
Comment (MahmoudNada0) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:371]: more details, **importBootDiskImage** function call **oi.buildBootDiskImageImportRequest** function to construct the request, it requires creating the data disks first becasue dataDisks are part of the request, it got initialized [here](https://github.com/GoogleCloudPlatform/compute-image-tools/pull/1864/files#diff-1ef6e3baf6905b5deba922e96bde4ae98152db2bce8590088b7d3e33add4695eR547)

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go (hunk 15 above)
Comment (MahmoudNada0) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go:427]: We create disks for data disks only, and create image for boot disk, and use these data disks to translate the boot-disk image, then the boot disk will be created from this image, it will be created on the final instance, and the created data disks will be attached to the final instance.

File: cli_tools/common/image/importer/bootable_disk_processor.go [hunk 23]
```diff
@@ -87,6 +92,25 @@ func newBootableDiskProcessor(request ImageImportRequest, wfPath string, logger
...
+		if wf.Steps["translate-disk"] != nil {
+			disks = &wf.Steps["translate-disk"].IncludeWorkflow.Workflow.Steps["translate-disk-inst"].CreateInstances.Instances[0].Disks
+		} else if wf.Steps["translate-disk-inst"] != nil {
+			// case translation wf that doesn't include another workflow (e.g. opensuse_15)
+			disks = &wf.Steps["translate-disk-inst"].CreateInstances.Instances[0].Disks
+		}
+
+		for _, dataDisk := range request.DataDisks {
+			*disks = append(*disks, &compute.AttachedDisk{Source: dataDisk.GetURI()})
+		}
```
Comment (EricEdens) [inline cli_tools/common/image/importer/bootable_disk_processor.go:110]: I'd suggest a follow-up PR that traverses the workflow as a tree and looks for `CreateInstances` nodes, rather than hardcoding the step names. I think that'll be more robust, since this code assumes naming conventions, which would cause the traversal to break.

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer_test.go [hunk 24]
```diff
@@ -549,14 +614,29 @@ type mockConfiguration struct {
...
+
+	mockComputeClient.EXPECT().CreateDisk(*params.Project, params.Zone, gomock.Any()).Return(nil).AnyTimes()
+
+	var imageName string
+	if params.InstanceNames == "" {
+		imageName = params.MachineImageName
+	} else {
+		imageName = params.InstanceNames
+	}
+	imageName += "-boot-image"
```
Comment (yswe) [inline cli_tools/gce_ovf_import/ovf_importer/ovf_importer_test.go:638]: I think `if` is not required here, you will always have `InstanceNames` specified due to the definition part.

File: cli_tools/gce_ovf_import/multi_disk_importer/request_executor_test.go [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+)
+
+const (
+	defaultProject = "project123"
+	defaultZone    = "zone1"
+)
+
+func TestExecuteRequests_HappyCase(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply cli_tools/gce_ovf_import/multi_disk_importer/request_executor_test.go:38]: I followed the same naming convention used in other test files.

File: cli_tools/gce_ovf_import/ovf_importer/ovf_importer_test.go [hunk 26]
```diff
@@ -549,14 +614,29 @@ type mockConfiguration struct {
 	descriptorFilenames []string
 	fileURIs            []string
 	images              []domain.Image
+	disks               []domain.Disk
 }
 
 func setupMocksAndRun(mockCtrl *gomock.Controller, params *ovfdomain.OVFImportParams, wfPath string, descriptor *ovf.Envelope, mockConfig mockConfiguration) (daisyutils.DaisyWorker, error) {
+	params.InstanceNames = strings.ToLower(strings.TrimSpace(params.InstanceNames))
+	params.MachineImageName = strings.ToLower(strings.TrimSpace(params.MachineImageName))
```
[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply cli_tools/gce_ovf_import/ovf_importer/ovf_importer_test.go:622]: Here I'm changing the image/instance name to lowercase, i'm not initializing it here.

File: cli_tools_tests/e2e/gce_ovf_import/test_suites/ovf_instance_import/ovf_instance_import_test_suite.go [hunk 27]
```diff
@@ -170,31 +177,21 @@ func deleteImagesIfInstanceCreationFails(ctx context.Context, testCase *junitxml
...
+		e2e.Failure(testCase, logger, fmt.Sprintf("Expected image %s to be removed", imgName))
+	}
+
+	// check if data disks are deleted
+	for i := 0; i < 2; i++ {
+		diskName := daisyovfutils.GenerateDataDiskName(props.instanceName, i)
+		disk, _ := client.GetDisk(testProjectConfig.TestProjectID, testProjectConfig.TestZone, diskName)
+		if disk != nil {
+			e2e.Failure(testCase, logger, fmt.Sprintf("Expected disk %s to be removed", diskName))
 		}
```
[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply cli_tools_tests/e2e/gce_ovf_import/test_suites/ovf_instance_import/ovf_instance_import_test_suite.go:195]: You're right, it's a mistake it should be i+1 instead of i, but it pass as it's search for disk-0 & disk-1 and don't don't find it which is expected. but I'll fix it to be i + 1.
regarding the boot image, I already checked the deletion of it, check line 182

[the remark this answers is not in the document]
Comment (yswe) [reply cli_tools_tests/e2e/gce_ovf_import/test_suites/ovf_instance_import/ovf_instance_import_test_suite.go]: Thank you for fixing!

Regarding the boot image, I saw that you checked for the deletion of the image, I am just curious if we need to check for the deletion of the boot **disk**. What do you think?

[the remark this answers is not in the document]
Comment (MahmoudNada0) [reply cli_tools_tests/e2e/gce_ovf_import/test_suites/ovf_instance_import/ovf_instance_import_test_suite.go]: in OVF-Import, the boot disk is created on the final instance, and in this test the final instance will not be created as the machine type is not correct.

[the remark this answers is not in the document]
Comment (yswe) [reply cli_tools_tests/e2e/gce_ovf_import/test_suites/ovf_instance_import/ovf_instance_import_test_suite.go]: It's exactly what I ask, got it, thank you for clarifying!
